whole-tale/ngx-dashboard
# fix: various fixes for the Run > Files view

## Problem
Fixes #89 - Files upload twice
Fixes #50 - Adding External Data would return a 400
Fixes #48 - Uploading a file would result in a 400, even though the file appeared to upload successfully
Fixes #41 - Create Folder in Home actually creates the folder in the Tale Workspace
Fixes #37 - Files cannot be removed/downloaded until after refreshing
Fixes #33 - Data registration dropdown should auto-select first result
Fixes #31 - Show previously-added data on select-data-modal
Fixes #30 - Add button should be disabled for read-only Tales
Fixes #28 - External Data shows nothing even after adding a dataset

## Approach
* Remove files from the upload queue after they complete
* Remove extra call to Girder's `/chunk` endpoint - logic for `initUpload` appears to upload correctly (#48)
* Fix `createFolder` to use the same logic for choosing `parentId` as the `uploadFiles` function uses (#41)
* Allow files to be removed and/or downloaded without needing to first refresh the view (#37)
* Auto-select first dropdown result when registering data (#33)
* Rework `select-data-modal` to show Selected Data on the right, as was previously true in the Ember.js dashboard (#31)
* Disable Add button on Run > Files for read-only Tales (#30) 
* Fixed display/fetch of External Data (#28)
* Remove extraneous debug line from Run > Interact view

## How to Test
Prerequisites: at least one Tale created, LIGO Tale

1. Checkout this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Choose a Tale you own and navigate to the Run > Files view
4. Select Home and upload a file
    * You should see that your file is uploaded successfully (#48)
    * You should see that your file has a Last Modified date (this indicates that it is now displaying an `Item`, as expected, and not a `File`)
5. Without refreshing the view, expand the dropdown beside the new file and choose Remove
    * The file should be removed from view (#37)
6. Select Home, create a folder, and refresh the page
    * You should still see the Home nav is still selected
    * Your new folder should still exist (#41)
7. Select External Data, click (+), and choose "Web (DOI or URL)"
    * The data registration modal should open
8. Enter `10.5065/D6862DM8` and search
    * You should see the Humans & Hydrology dataset is found
    * The dataset should be automatically selected in the dropdown (#33)
9. Click Register to close the modal, then click (+) again, and choose "WT Data Catalog"
     * The Select Data modal should open
10. Click the "Humans & Hydrology" dataset to select it, optionally double-click to drill down into it and select individual folders/files, and click "Add Selected. and then 
    * The modal should close
    * You should see the new dataset(s) appear on the right after adding it
11. Click "Save" below to confirm
    * You should see the new dataset(s) appear in the view shortly after the modal closes
    * You should no longer see 400 errors appear in the console when adding the dataset (#50)
12. Click (+) again, and choose "WT Data Catalog" once more
    * You should see the data you've previously added appears on the right by default (#31)
13. Refresh the page
    * You should still see the External Data nav is still selected
    * Your new dataset should still exist on the Tale (#28)
14. Navigate to the Run > Files view for the LIGO Tale
    * You should see the (+) button at the top-right is disabled (#30)
